File: Python/src/bin/easy_189.py
# Daily Coding Problem: 189
# Difficulty: Easy
# Asked by: Google
# Authored by: TenthEdict

def longest_distinct_subarray(arr):
    positions = {}
    beginning = 0
    result = 0

    for i in range(len(arr)):
        element = arr[i]

        if element in positions and positions[element] >= beginning:
            beginning = positions[element] + 1

        positions[element] = i

        result = max(result, i - beginning + 1)

    return result

print(longest_distinct_subarray([1,2,3,2,4, 5])) # returns 4
print(longest_distinct_subarray([5, 1, 3, 5, 2, 3, 4, 1])) # returns 5
print(longest_distinct_subarray([1,1,1,1])) # returns 1
print(longest_distinct_subarray([1,2,3,4,5])) # returns 5
print(longest_distinct_subarray([5,4,3,2,1,1])) # returns 5
print(longest_distinct_subarray([])) # returns 0

File: Rust/src/bin/easy_189.rs
// Daily Coding Problem: 189
// Difficulty: Easy
// Asked by: Google
// Authored by: TenthEdict

use std::collections::HashMap;

fn longest_distinct_subarray(arr: &[i32]) -> usize {
    let mut positions: HashMap<i32, usize> = HashMap::new();
    let mut beginning: usize = 0;
    let mut result: usize = 0;

    for (i, e) in arr.iter().enumerate() {
        if let Some(old_pos) = positions.insert(*e, i) {
            if old_pos >= beginning {
                beginning = old_pos + 1;
            }
        }
        result = result.max(i - beginning + 1);
    }
    result
}

fn main() {
    println!("{}", longest_distinct_subarray(&[1, 2, 3, 2, 4, 5])); // 4
    println!("{}", longest_distinct_subarray(&[5, 1, 3, 5, 2, 3, 4, 1])); // 5
    println!("{}", longest_distinct_subarray(&[1, 1, 1, 1])); // 1
    println!("{}", longest_distinct_subarray(&[1, 2, 3, 4, 5])); // 5
    println!("{}", longest_distinct_subarray(&[])); // 0
}
